[mkFit] Propagate to plane -- technical changes to support further development

osschar opened the pull request:
#### PR description:

Implement propagation to plane and Kalman update to plane -- work done by @cerati 

Propagation to plane is used for propagation to the hit, while propagation to r and z are used for propagation to the layer center. Propagation to the hit starts from the state at the previous hit (by re-calling the `mkfndr->inputTracksAndHitIdx(...)` ).

This also updates the material effects, as developed by @slava77.

Both changes are currently disabled by default via `mkfit::Config` switches, making this a technical PR to support further development, especially for layer-processing optimizations & improvements and for supporting Phase2 tilted geometry.


#### PR validation:

MTV plots in preparation, there should be no significant changes.

cerati committed:
[mkFit] Implement vectorized propagate-to-plane and Kalman update on a plane.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43146/37427

ERROR: Build errors found during clang-tidy run.
```
RecoTracker/MkFitCore/src/PropagationMPlex.icc:70:3: error: no matching function for call to 'parsFromPathL_impl' [clang-diagnostic-error]
  parsFromPathL_impl(inPar, outPar, kinv, s, nmin, nmax);
  ^~~~~~~~~~~~~~~~~~
RecoTracker/MkFitCore/src/PropagationMPlex.icc:418:3: note: in instantiation of function template specialization 'parsAndErrPropFromPathL_impl<Matriplex::Matriplex<float, 6, 1, 4>, Matriplex::Matriplex<int, 1, 1, 4>, Matriplex::Matriplex<float, 6, 1, 4>, Matriplex::Matriplex<float, 6, 6, 4>, Matriplex::Matriplex<float, 1, 1, 4>>' requested here
--
RecoTracker/MkFitCore/src/PropagationMPlex.icc:383:5: error: no matching function for call to 'parsFromPathL_impl' [clang-diagnostic-error]
    parsFromPathL_impl(inPar, outParTmp, kinv, s, nmin, nmax);
    ^~~~~~~~~~~~~~~~~~
RecoTracker/MkFitCore/src/PropagationMPlex.cc:1094:5: note: in instantiation of function template specialization 'helixAtPlane_impl<Matriplex::Matriplex<float, 6, 1, 4>, Matriplex::Matriplex<int, 1, 1, 4>, Matriplex::Matriplex<float, 6, 1, 4>, Matriplex::Matriplex<float, 3, 1, 4>, Matriplex::Matriplex<float, 6, 6, 4>, Matriplex::Matriplex<float, 1, 1, 4>>' requested here
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43146/37428

- This PR adds an extra 176KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc  modified in PR(s): #43145
  - File  RecoTracker/MkFitCore/interface/Config.h  modified in PR(s): #43145
  - File  RecoTracker/MkFitCore/interface/TrackerInfo.h  modified in PR(s): #43145
  - File  RecoTracker/MkFitCore/src/Matrix.h  modified in PR(s): #43145
  - File  RecoTracker/MkFitCore/src/MkBuilder.cc  modified in PR(s): #43145
  - File  RecoTracker/MkFitCore/src/MkFinder.cc  modified in PR(s): #43145
  - File  RecoTracker/MkFitCore/src/MkFinder.h  modified in PR(s): #43145

cmsbuild commented:
A new Pull Request was created by  @osschar (Matevž Tadel) for master.

It involves the following packages:

- RecoTracker/MkFit (**reconstruction**)
- RecoTracker/MkFitCore (**reconstruction**)


@mandrenguyen, @cmsbuild, @jfernan2 can you please review it and eventually sign? Thanks.
@JanFSchulte, @mtosi, @VourMa, @rovere, @felicepantaleo, @gpetruc, @VinInn, @missirol, @GiacomoSguazzoni, @dgulhan, @makortel, @mmusich this is something you requested to watch as well.
@rappoccio, @antoniovilela, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

osschar commented:
There is a minimal conflic between this PR and PR # #43145.
It occurs around `MkFitCore/src/MkFinder.cc:170` -- both changes should be accepted.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35490/summary.html
**COMMIT**: 03da3737b7b13488f995cf31d99b4e21045a3607
**CMSSW**: CMSSW_13_3_X_2023-10-29-0000/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43146/35490/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 37 lines to the logs
 - Reco comparison results: 18293 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3362691
 - DQMHistoTests: Total failures: 20917
 - DQMHistoTests: Total nulls: 3
 - DQMHistoTests: Total successes: 3341749
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.043 KiB( 49 files compared)
 - DQMHistoSizes: changed ( 141.044 ): 0.043 KiB        JetMET/SUSYDQM
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-29-0000+e36544/59481/triggerResults): no differences found

mandrenguyen commented on RecoTracker/MkFitCore/src/PropagationMPlex.cc:
```diff
@@ -488,7 +488,19 @@ namespace mkfit {
 
+    //helixAtRFromIterativeCCS_impl_new(inPar, inChg, msRad, outPar, errorProp, outFailFlag, 0, NN, N_proc, pflags);
+    helixAtRFromIterativeCCS_impl(inPar, inChg, msRad, outPar, errorProp, outFailFlag, 0, NN, N_proc, pflags);
+    /*
```
Is there a good reason to keep all this commented code?

osschar replied:
I have removed this (and the much longer implementation in the .icc file) for now and left a comment below  with a reference to specific commit in case this needs to be revisited.

leonardogiannini commented:
the last commits are needed to remove (almost) all the differences in the code that is run by default, making this effectively a technical PR, as already stated.

MTV comparisons here
http://uaf-10.t2.ucsd.edu/~legianni/plots_compare_propToPlane_PRVersion6/

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43146/37511

- This PR adds an extra 196KB to repository

cmsbuild commented:
Pull request #43146 was updated. @cmsbuild, @jfernan2, @mandrenguyen can you please check and sign again.

slava77 commented:
> Conflicting files  RecoTracker/MkFitCore/src/MkFinder.cc 

@osschar @leonardogiannini 
please check and rebase or resolve the conflict.
Thank you.

smuzaffar commented:
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

osschar commented:
Rebased to head of CMSSW_13_3_X (currently still == master) and forced pushed to source branch.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43146/37531

- This PR adds an extra 148KB to repository

cmsbuild commented:
Pull request #43146 was updated. @cmsbuild, @jfernan2, @mandrenguyen can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35647/summary.html
**COMMIT**: bbfe0c29f8ae9727335e0428628b4f329a08a41b
**CMSSW**: CMSSW_13_3_X_2023-11-06-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43146/35647/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @smuzaffar cms-sw/cmssw#43170
 - @fabiocos cms-sw/cmssw#43124
 - @bsunanda cms-sw/cmssw#43167
 - @bsunanda cms-sw/cmssw#43194

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35647/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35647/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially added 217 lines to the logs
 - Reco comparison results: 5666 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3363010
 - DQMHistoTests: Total failures: 27363
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3335625
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-11-06-1100+e36544/59594/triggerResults): no differences found

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35728/summary.html
**COMMIT**: bbfe0c29f8ae9727335e0428628b4f329a08a41b
**CMSSW**: CMSSW_14_0_X_2023-11-09-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43146/35728/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 355 lines from the logs
 - Reco comparison results: 5757 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3363010
 - DQMHistoTests: Total failures: 87098
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3275890
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-09-1100+e36544/59656/triggerResults): no differences found

osschar commented:
I have now removed all the commented-out new code. There were two main segments, reimplementations of prop-to-Z and prop-to-R with separated parameter and error propagation (as it is also implemented in the new prop-to-plane).

@cerati, just in case, here are the links to commits where this functions have been removed:
- helixAtZ_new https://github.com/cms-sw/cmssw/pull/43146/commits/2488a10caaa45679cd4f8c4ac650706491a98602
- helixAtRFromIterativeCCS_impl_new https://github.com/cms-sw/cmssw/pull/43146/commits/c9b4d5ef191a1f3a921f91032d866fe590873723

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43146/37627

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #43146 was updated. @cmsbuild, @mandrenguyen, @jfernan2 can you please check and sign again.

mandrenguyen commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35758/summary.html
**COMMIT**: c9b4d5ef191a1f3a921f91032d866fe590873723
**CMSSW**: CMSSW_14_0_X_2023-11-12-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43146/35758/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 35 lines from the logs
 - Reco comparison results: 6212 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3363028
 - DQMHistoTests: Total failures: 26112
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3336894
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-12-2300+e36544/59688/triggerResults): no differences found

mandrenguyen commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @sextonkennedy, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)

slava77 commented:
@cms-sw/orp-l2 
can this be merged soon?
We have a small less technical followup that was excluded to simplify this large PR integration.

It would be nice to get that followup in pre1 scheduled for next week.
So, having this merged today/tomorrow is important.
Thank you.

antoniovilela commented:
@osschar @slava77 
It is difficult to understand what has been addressed in the PR from the commit history.
Could you squash the commits in groups related to different issues, with meaningful commit messages. Or simply squash all commits into one?
Thanks.

osschar commented:
Yes, good point, I will rebase to a single commit.

osschar commented:
Rebase done.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43146/37764

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #43146 was updated. @mandrenguyen, @cmsbuild, @jfernan2 can you please check and sign again.

antoniovilela commented:
> Rebase done.

Thanks. Let's refresh the tests and then we merge.

antoniovilela commented:
please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35919/summary.html
**COMMIT**: 7fab695247933f6f66a9791ca94e3b6b6bb5d88b
**CMSSW**: CMSSW_14_0_X_2023-11-17-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43146/35919/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

- [136.881](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35919/runTheMatrixINPUT-results/136.881_RunMuOnia2018C)```136.881_RunMuOnia2018C/step2_RunMuOnia2018C.log```


## Comparison Summary



**Summary**:
 - You potentially added 281 lines to the logs
 - Reco comparison results: 6204 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3363070
 - DQMHistoTests: Total failures: 26110
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3336938
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-17-1100+e36544/59820/triggerResults): no differences found

antoniovilela commented:
please test workflow 136.881

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e36544/35934/summary.html
**COMMIT**: 7fab695247933f6f66a9791ca94e3b6b6bb5d88b
**CMSSW**: CMSSW_14_0_X_2023-11-17-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43146/35934/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 360 lines to the logs
 - Reco comparison results: 6916 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3415195
 - DQMHistoTests: Total failures: 26115
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3389058
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 218 log files, 170 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-17-1100+e36544/59832/triggerResults): no differences found

antoniovilela commented:
+1

antoniovilela commented:
merge
   - @cms-sw/reconstruction-l2 signed before squash.

perrotta commented:
(Unsolicited comment, please ignore if you think so). I understand the urgency to merge this PR. However, given it touches very important areas of tracking I think that the code should be kept cleaner, easier to investigate and worked on, if needed. There are large chunks of commented out code that would at least deserve some comment: in particular, [the last rebase](https://github.com/cms-sw/cmssw/compare/c9b4d5ef191a1f3a921f91032d866fe590873723..7fab695247933f6f66a9791ca94e3b6b6bb5d88b) only brought in commented out lines of codes. Another thing I noticed: in PropagationMPlex the `DEBUG` sections may not work, because the definition of the `debug` variable (at least) is also commented out. Did you ever try it? I suspect that some extra gymnastic would be needed to allow debugging, besides just switching the `DEBUG` flag at line https://github.com/cms-sw/cmssw/blob/7fab695247933f6f66a9791ca94e3b6b6bb5d88b/RecoTracker/MkFitCore/src/PropagationMPlex.cc#L8

antoniovilela commented:
> (Unsolicited comment, please ignore if you think so). I understand the urgency to merge this PR. However, given it touches very important areas of tracking I think that the code should be kept cleaner, easier to investigate and worked on, if needed. There are large chunks of commented out code that would at least deserve some comment: in particular, [the last rebase](https://github.com/cms-sw/cmssw/compare/c9b4d5ef191a1f3a921f91032d866fe590873723..7fab695247933f6f66a9791ca94e3b6b6bb5d88b) only brought in commented out lines of codes. Another thing I noticed: in PropagationMPlex the `DEBUG` sections may not work, because the definition of the `debug` variable (at least) is also commented out. Did you ever try it? I suspect that some extra gymnastic would be needed to allow debugging, besides just switching the `DEBUG` flag at line
> 
> https://github.com/cms-sw/cmssw/blob/7fab695247933f6f66a9791ca94e3b6b6bb5d88b/RecoTracker/MkFitCore/src/PropagationMPlex.cc#L8

Thanks Andrea.
@osschar @slava77 
Matevž, could you follow up?

osschar commented:
Indeed, the last commit from before the rebase, https://github.com/cms-sw/cmssw/commit/c9b4d5ef191a1f3a921f91032d866fe590873723, has been lost. I must have made a mistake, I apologize for this. How do you want me to handle this, open a new PR with just this change?

The DEBUG thing is like this by design -- it allows us to switch debug printouts on/off with finer granularity than on the level of whole files.
https://github.com/cms-sw/cmssw/blob/master/RecoTracker/MkFitCore/src/Debug.h

antoniovilela commented:
> Indeed, the last commit from before the rebase, [c9b4d5e](https://github.com/cms-sw/cmssw/commit/c9b4d5ef191a1f3a921f91032d866fe590873723), has been lost. I must have made a mistake, I apologize for this. How do you want me to handle this, open a new PR with just this change?
> 
> The DEBUG thing is like this by design -- it allows us to switch debug printouts on/off with finer granularity than on the level of whole files. https://github.com/cms-sw/cmssw/blob/master/RecoTracker/MkFitCore/src/Debug.h

Hi Matevž,
Yes, you can open a new PR. In the description mention it is a follow up of this.
To squash the commits into one, you could use the git cms-squash-topic command, introduced recently (https://cms-sw.github.io/tutorial-resolve-conflicts.html).
Thanks.

jfernan2 commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.

osschar commented:
Thank you @antoniovilela! I have opened a follow up PR, #43356.

antoniovilela commented:
> Thank you @antoniovilela! I have opened a follow up PR, #43356.

Thanks
